### PR TITLE
Add Dart support in interop test suite

### PR DESCRIPTION
**Pull request by jakobr-google**

### CODE REVIEW

File: templates/tools/dockerfile/interoptest/grpc_interop_dart/build_interop.sh.template [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  set -e
+
+  mkdir -p /var/local/git
+  git clone /var/local/jenkins/grpc-dart /var/local/git/grpc-dart
+
+  # copy service account keys if available
+  cp -r /var/local/jenkins/service_account $HOME || true
+
+  cd /var/local/git/grpc-dart/interop
+  /usr/lib/dart/bin/pub get
```
[the remark this answers is not in the document]
Comment (jakobr-google) [reply templates/tools/dockerfile/interoptest/grpc_interop_dart/build_interop.sh.template]: It uses the code from the checked out repo, yes. Interop’s pubspec.yaml has a path dependency for grpc: https://github.com/dart-lang/grpc-dart/blob/master/interop/pubspec.yaml

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: tools/interop_matrix/client_matrix.py [hunk 2]
```diff
@@ -18,6 +18,7 @@
 
 def get_github_repo(lang):
     return {
+        'dart': 'git@github.com:dart-lang/grpc-dart.git',
```
Comment (adelez) [inline tools/interop_matrix/client_matrix.py]: Is there a release to test? If not, it's better to add this when there is one.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: tools/interop_matrix/client_matrix.py (hunk 2 above)
Comment (jakobr-google) [reply tools/interop_matrix/client_matrix.py]: Define release? We're currently at v0.2.1 on Pub: https://pub.dartlang.org/packages/grpc.

**Comment by jakobr-google** [comment]
I've signed up with CNCF. Dart gRPC does have auth support, but I hadn't implemented the interop tests for it yet. I'll wire them up early next week when I'm back from DartConf.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by jtattermusch** [comment]
I re-ran the tests, and most auth tests have indeed been fixed by https://github.com/dart-lang/grpc-dart/pull/55.

The only remaining test is now `compute_engine_creds`

```
Failure
Unhandled exception:
Service account key file not specified.
#0      Tester._readServiceAccountJson (package:interop/src/client.dart:49:7)
#1      Tester.serviceAccountJson (package:interop/src/client.dart:45:31)
#2      Tester.computeEngineCreds (package:interop/src/client.dart:673:10)
<asynchronous suspension>
#3      Tester.runTestCase (package:interop/src/client.dart:137:16)
<asynchronous suspension>
#4      Tester.runTest (package:interop/src/client.dart:108:11)
<asynchronous suspension>
#5      main (file:///var/local/git/grpc-dart/interop/bin/client.dart:107:20)
<asynchronous suspension>
#6      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:263)
#7      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:151)
```

https://sponge.corp.google.com/target?id=1e713124-8173-4afb-b512-5f4cec4e0c50&target=github/grpc&searchFor=&show=ALL&sortBy=STATUS

**Comment by jakobr-google** [comment]
I think I've fixed the compute_engine_creds test. Could you kick off a new run to verify, please?

**Comment by jtattermusch** [comment]
@jakobr-google, started the run.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by jtattermusch** [comment]
Its failing with an unrelated node problem:
```
npm ERR! sha512-bqB1yS6o9TNA9ZC/MJxM0FZzPnZdtHj0xWK/IZ5khzVqdpGul/R/EIiHRgFXlwTD7PSIaYVnGKq1QgMCu2mnqw== integrity checksum failed when using sha512: wanted sha512-bqB1yS6o9TNA9ZC/MJxM0FZzPnZdtHj0xWK/IZ5khzVqdpGul/R/EIiHRgFXlwTD7PSIaYVnGKq1QgMCu2mnqw== but got sha512-OvqKfUUBKvxc/W6hoiKPVgytZR/5B1FyP22xUo41Hq26Qf7Qn3G4SOMJa6Ez5fsPwbHszOvGrNyUJw8jGSs7BQ==. (1081344 bytes)

npm ERR! A complete log of this run can be found in:
npm ERR!     /tmp/npm-cache/_logs/2018-02-07T08_34_56_247Z-debug.log
+ EXITCODE=1
+ docker rm -f build_grpc_interop_node_0b386b04-14f6-432e-8059-fc23f3593ff2
build_grpc_interop_node_0b386b04-14f6-432e-8059-fc23f3593ff2
+ exit 1

2018-02-07 00:35:06,015 FAILED: build_docker_node [ret=1, pid=9075, time=238.3sec]
```

which prevents the tests from running. I'm looking into this.

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_pump.BM_PumpStreamServerToClient_MinSockPair__0.opt.old: 1


[microbenchmarks] No significant performance differences
```

**Comment by jtattermusch** [comment]
To work around the Node.js docker build problem, I've setup a separate PR just to test this:
https://github.com/grpc/grpc/pull/14352

and the results show that `compute_engine_creds` is now passing.

https://sponge.corp.google.com/target?id=88f4fd4b-d51f-47fb-b9e7-1947e411ca30&target=github/grpc&searchFor=&show=ALL&sortBy=STATUS

Thanks for the fix!

**Comment by jtattermusch** [comment]
I'm now happy with the state of interop tests, so here are the next steps:
- I'll initiate the transition to grpc/grpc-dart internally (see e-mail).
- Once the repository has been transitioned, we can adjust this PR to point to the new repository and merge it (there's no point of merging now, then transitioning which would break everything and fixing in a separate PR)

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by jtattermusch** [comment]
@jakobr-google  I started the tests for the latest changes, once the test finish everything is green, I think we can proceed with a merge.

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by jakobr-google** [comment]
@jtattermusch looks like there's an issue with building the Dart interop Docker image:
```
+ BASE_IMAGE=grpctesting/grpc_interop_dart_87b2648b79c2eaff5d39d5cd71b7b069ce3c883d
+ docker pull grpctesting/grpc_interop_dart_87b2648b79c2eaff5d39d5cd71b7b069ce3c883d
Using default tag: latest
Error response from daemon: pull access denied for grpctesting/grpc_interop_dart_87b2648b79c2eaff5d39d5cd71b7b069ce3c883d, repository does not exist or may require 'docker login'
```

Is there something I need to add?

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by jtattermusch** [comment]
@jakobr-google  FYI when the base Dockerfiles are updated, the new dockerimages need to be pushed to dockerregistry using (https://github.com/grpc/grpc/blob/master/tools/dockerfile/push_testing_images.sh). I just did that.

**Comment by jtattermusch** [comment]
Ad interop_to_prod failure:
There is currently a breakage for all interop_to_prod tests using gateway_v4  that's unrelated to this PR. 
Other dart interop tests are passing (including a sufficient subset of cloud_to_prod tests) and they've been also been passing before the interop test breakage.

**Pull request merged**